econml/sklearn_extensions/selective_regularization.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""Provides linear regressors with support for applying L1 and/or L2 regularization to a subset of coefficients."""
+
+from warnings import warn
+import numpy as np
+from sklearn import clone
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import r2_score
+
+
+class SelectiveRegularization:
+    """
+    Estimator of a linear model where regularization is applied to only a subset of the coefficients.
+
+    Assume that our loss is
+
+    .. math::
+        \\ell(\\beta_1, \\beta_2) = \\lVert y - X_1 \\beta_1 - X_2 \\beta_2 \\rVert^2 + f(\\beta_2)
+
+    so that we're regularizing only the coefficients in :math:`\\beta_2`.
+
+    Then, since :math:`\\beta_1` doesn't appear in the penalty, the problem of finding :math:`\\beta_1` to minimize the
+    loss once :math:`\\beta_2` is known reduces to just a normal OLS regression, so that:
+
+    .. math::
+        \\beta_1 = (X_1^\\top X_1)^{-1}X_1^\\top(y - X_2 \\beta_2)
+
+    Plugging this into the loss, we obtain
+
+    .. math::
+
+        ~& \\lVert y - X_1 (X_1^\\top X_1)^{-1}X_1^\\top(y - X_2 \\beta_2) - X_2 \\beta_2 \\rVert^2 + f(\\beta_2) \\\\
+        =~& \\lVert (I - X_1 (X_1^\\top  X_1)^{-1}X_1^\\top)(y - X_2 \\beta_2) \\rVert^2 + f(\\beta_2)
+
+    But, letting :math:`M_{X_1} = I - X_1 (X_1^\\top  X_1)^{-1}X_1^\\top`, we see that this is
+
+    .. math::
+        \\lVert (M_{X_1} y) - (M_{X_1} X_2) \\beta_2 \\rVert^2 + f(\\beta_2)
+
+    so finding the minimizing :math:`\\beta_2` can be done by regressing :math:`M_{X_1} y` on :math:`M_{X_1} X_2` using
+    the penalized regression method incorporating :math:`f`.  Note that these are just the residual values of :math:`y`
+    and :math:`X_2` when regressed on :math:`X_1` using OLS.
+    """
+
+    def __init__(self, unpenalized_inds, penalized_model, fit_intercept=False):
+        self._unpenalized_inds = unpenalized_inds
+        self._penalized_model = clone(penalized_model)
+        self._fit_intercept = fit_intercept
+
+    def fit(self, X, y, sample_weight=None):
+        self._penalized_inds = np.delete(np.arange(X.shape[1]), self._unpenalized_inds)
```

Comment:
We most probably want to add check_inputs here (using for instance the sklearn check_inputs).

Reply:
Now uses `check_X_y`